### ISSUE TITLE
Document the steps to make a sim accessible

### ISSUE THREAD

**Issue by zepumph** [phetsims/a11y-research#22]
It would be good to have a central document, similar to 'how-to-instrument-a-simulation-for-phet-io.md` (but perhaps less verbose) about the steps needed to outfit a sim. This process will take time to refine and iterate, so it's good to get started now. @jessegreenberg and I are starting from the beginning with build-an-atom, so this seems like a great place to start.

**Comment by terracoda** [comment]
@zepumph, the research articles (HCII 2016, Assets2016 and CSUN 2017) from the PhET research may be a helpful.

**Comment by zepumph** [comment]
@terracoda, would you mind including links to these articles in this issue. I was unable to find these on PhET Research or PhET Accessibility Research, sorry about that.

**Comment by terracoda** [comment]
@zepumph, sorry I did not tell the whole truth: 
- To get to the HCII paper, you need to go through the library, it is not in an open access journal.
    - A Balloon, a Sweater, and a Wall: Developing Design Strategies for Accessible User Experiences with a Science Simulation. Here's the DOI: 0.1007/978-3-319-40238-3_15
- The PDF of ASSETS poster is here: [Demonstration: Screen reader support for a complex interactive science simulation](https://drive.google.com/file/d/0B44Uycdx6JGdRFpXcDJqZl9BUk0/view)
- The CSUN paper is in an open access journal, but not yet available. It should be available soon.

**Comment by terracoda** [comment]
@zepumph and @jessegreenberg, here's a nice article on _[What is an accessible name?](https://www.paciellogroup.com/blog/2017/04/what-is-an-accessible-name/)_ by Léonie Watson. It has some nice info about how names are added to the Accessibility Tree.

**Comment by zepumph** [comment]
From a team meeting today, @jessegreenberg we should find a time to make this document happen for general dev ingestion.

**Comment by zepumph** [comment]
phetsims/scenery#686 and phetsims/scenery#748 have been done. We should start on this. Marking as high priority so I can get a start on this this week.

**Comment by zepumph** [comment]
I'm worried about putting this out to everyone right now, because it's basically an outline right now, and we may discuss much more than is helpful (since it is in such an early state). Let's try to limit feedback to outline like comments. I want to know what you think about structure. 
* What is the purpose of this document. Is it an intro to the dev side of a11y, or is it the complete how to on everything a11y, just written in a non overwhelming way, so you can take it chunk by chunk. The phet-io doc did it in the latter manner.
* Any additions to categories? 
* Things you would replace? 
* What about flow, think about how someone would want to learn this stuff, in order (keyboard nav --> static description --> alerts).
* I tried to outline where each mainstream a11y option should be talked about. Let me know if:
    * they are in weird spots
    * they are too complicated of options, and not worth mentioning. i.e. `focusHighlightLayerable` when you are just learning what the heck focus is?!?!? I must be crazy this Friday afternoon!

@jessegreenberg @mbarlow12 @emily-phet @terracoda. Assigning everyone.

**Comment by emily-phet** [comment]
@zepumph First, this looks like a great starting point!

Some thoughts:
- The goal of this doc would be to support developers in getting up to speed on a11y. We'll work on other documentation (e.g., design documentation and likely 'style guide' style specs on specific form elements, etc.) to support the designers and areas where designers and devs overlap.
- The primary goals of this document will shift a bit over time. For now, it will be used to help people who already know a lot about sim development understand how to incorporate a11y. Down the road, it will be used when people are learning how to do sim development (which will include a11y). Perhaps having some overview/summary pieces, and then broken out categories for specific areas of sim development may help bridge these two purposes. This may be what you've done, I'm probably not following some of the flow well - not being a dev.
- Also, you can link to the paper on the PDOM architecture. That will provide an intro to what the PDOM is, and why it's needed. It will be published officially soon, but you can use our pre-print in the meantime.
- As we found with creating static descriptions, it may not actually work well to try to provide a lot of documentation in advance. I think you're off to a good start here creating a framework to be added on to. We'll really just need to get it to a place where others can understand *enough to critique, suggest, and edit themselves* rather than trying to create something totally understandable and actionable out of the gate. 
- Also, not biting off too much right now is good, but looking forward, sonification will be included and won't necessarily have a relationship to the pdom the way the navigation and descriptive content does...so more sections for this kind of feature will be needed - later on!

**Comment by jessegreenberg** [comment]
We reviewed this at dev meeting, @jessegreenberg and @mbarlow12 will review this and comment.

**Comment by terracoda** [comment]
@zepumph, I do not know what a DAG feature is.

I added `label` tot eh accessible name section, and worked a bit on the google doc referenced in MD doc.

**Comment by terracoda** [comment]
@zepumph, I added a section at the beginning, " Understanding the Accessibility Tree". I think this is an important concept that developers need to understand. 

The details I added in the Accessible Name section might be too detailed at this point. The concepts of Accessibility Tree and Accessible Name. An accessible name has to get to the Accessibility Tree in order to reach the assistive technology.

**Comment by zepumph** [comment]
@terracoda's commits (from oldest to newest):

5f53b206da8aa0bb0d0b53b8fb7ffe303e5c38ba
cc26c9e3c694810bf23380aa964b4ea3cce1b70b
fead4106b69651fa05df8ff4b13a01f893309cb2
c7a62f6bc03af73db8991fba5dc9fca42f0593b2

(FYI, although you probably already know this, if you add "#22", "phetsims/a11y-research#22", or "https://github.com/phetsims/a11y-research/issues/22" to the commit, it shows up in the issue)

**Comment by terracoda** [comment]
@zepumph, I didn't know about "#22" - thanks!

**Comment by zepumph** [comment]
I'm updating this today and tomorrow, no use reviewing the previous version above until I have a new rough draft!

**Comment by zepumph** [comment]
I took a new strategy today. @jonathanolson and I recently found that there are accessibility docs in scenery that need updating. I thought that I would springboard off of that work for this. As a result you should see a very different page than from before. It is now full HTML, not markdown. 

It can be viewed here http://phettest.colorado.edu/scenery/doc/accessibility.html, but note that it is not yet complete. It uses a built version of scenery to render examples into the page, and phettest doesn't have that build version yet, hopefully soon (if you see white rectangles, it is broken, if they are blue, it is working). I hope to get that working by tomorrow. @emily-phet, like you suggested I tried hard to hone in this document to focus on our developers, as a result it is a bit terse in places.

In general, I tried to have this doc be less of a traditional code doc, listing methods, and more explanation. Further exploration and implementation details is meant to be in the js code itself (I think).

**Comment by zepumph** [comment]
@mbarlow12 and @jessegreenberg, if you want to test locally, just `grunt` in scenery first.

**Comment by zepumph** [comment]
(It now works with live scenery)

**Comment by terracoda** [comment]
@zepumph, this document is looking really good. I am reading through it.

**Comment by terracoda** [comment]
@zepumph, here's some feedback. I've just read  up until the Dynamic Descriptions section. If you prefer that I make direct edits, just let me know.

**Regarding Remove Wall Example:**
I don't understand `labelContent`: 'Press the remove wall button below.'

- Note that nothing like this exists in the pDOM. In the case of the this example, I would assume that `innerContent` is the `labelContent`.

In the pDOM the label for the wall node is just ```<h3>Wall</h3>```. The remove wall button is actually within the Control panel.

In this example, the _Wall object_ in the Play Area and the _Remove Wall button_ in the Control Panel are separate elements in the pDOM and might also be separate nodes in the scene graph tree.

The _Remove Wall button_ has a description that is in fact help text.

The _Wall object_ has a description that is in fact the state of the wall.

**Section Input types (defaults)**
Does the input have a default type? 
If not, I think it should, or it should be required that a type be included.

**Typo in Input types section:**
accesibleChecked -> accessibleChecked

Correction, maybe?
- "listener is often different" -> listeners for the keyboard are always different, right? A mouse listener listens for clicks, keyboard listeners listen for specific key presses.

**Focus section**
- ToDo: add something about what actually gets focus.
- Do we need to say why?

All interactive elements in the pDOM receive keyboard focus. Not all sim objects are interactive. For example, the Sweater in BASE is a dynamic content onbject, but is not directly interacted with by the user, thus the sweater never receives a focus highlight. 

**Dynamic Descriptions section**
- ToDo: make sure it is clear the difference between dynamic state information and dynamic interactive alerts.

**Comment by zepumph** [comment]
I made all of your above suggestions except:
- [ ] Regarding Remove Wall Example. . . 
- [ ] Section Input types (defaults) (handled in https://github.com/phetsims/scenery/issues/776)

**Comment by zepumph** [comment]
After talking with @jessegreenberg and @mbarlow12, we have the following list to do before delivering to the dev team for iteration.

- [ ] Short master list of accessibile html resources to make every developer using the a11y api read prior to this doc (ideally):
  * https://developers.google.com/web/fundamentals/accessibility/semantics-builtin/the-accessibility-tree
  * https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML
- [ ] (a) Rename ". . . Further Understanding" header to "Next Steps for Understanding". (b) Add a sub header header to the list of phet's hyperlinks saying "PhET Published Resources" or something. (c) speak more to where you can get more information about the scenery accessibility inplementation (source code?)
- [ ] adapt the accessible name section. Assume that they have read the link in the top paragraph (https://developer.paciellogroup.com/blog/2017/04/what-is-an-accessible-name/) and then each bullet should explain how to accomplish that implementation with the accessibility api.
- [ ] swap out the "Remove Wall Example" for a basic "Magnet" example (may be easier to have label/description). Note that most interactives (all that we have encountered) probably won't need `innerContent` and `labelContent`.
- [ ] add another example after the Magnet that illustrates how all the options are "arbitrary" to a sense, and can just be used to create lots of text in one `Node`.
- [ ] Add another example after the text one that uses the scene graph to show how nesting occurs in the PDOM by default. (`<ul>` parent and a few `<li>` children or something (that can be nice for Radio Buttons)). Make this example use getters and setters to illustrate that you can do both.
- [ ] Move "overall code structure" section into it's own file called "Accessible Implementation Notes". This will be a small document, but it will be good to have a place for it.

**Comment by jessegreenberg** [comment]
- [ ] Add information  about  "accessible name" back into the "Manipulating the PDOM" section, with inline link about the accessible name for easy reference.
- [ ] Change example functions in "Manipulating the PDOM" section to be the ES5 getters/setters.
- [ ] Add a note to "PhET Specific" section that says that options are prefered to getters/setters.
- [ ] Widen  magnet example display.
- [ ] Make sure that terminology for sibling/parent is clear. These are **not** "siblings of the node". They are sibling elements for the accessible content (the primary DOM element).
- [ ] Specify a containerTagName in "multiple DOM Elements" section for clarity
- [ ] Add a section that describes "The structure of the PDOM".
  - [ ]  Move "Leveraging  the Scene Graph" example under this.
  - [ ] Clarify that accessible content for Node children will only be added as children to the primary DOM element. (under  "The structure ..." section)
  - [ ] Setting a tagName to null will clear it out (possibly not under structure section).

**Comment by zepumph** [comment]
This issue is complete in my book, we have delivered this file to other devs, and https://github.com/phetsims/scenery/issues/777 will be used to track changes from here. Closing. Please reopen if anyone else feels differently.
